### ISSUE TITLE
Leanfront mythtv commercial skipping

### ISSUE THREAD

**Issue by rocknrobin72** [bennettpeter/android-MythTV-Leanfront#38]
I recently used adb to sideload leanfront to my Sony Bravia Android T.V.  I am able to connect to my backend, in fact I have three of them and I can manually input the ip addresses to connect to all three, but I noticed a couple of drawbacks that may or may not be correctable.  

These issues will be pointed to mythtv ver. 32.  

Number one:  I have been using Kodi with the mythtv pvr client add-on and it worked great with mythtv ver.  31.  When I upgraded one of my backends to ver. 32 I noticed that most of the recorded programs were marked with yellow lettering which did indicate a damaged recording on ver. 31.  Now when I try to play the "supposedly marked damaged" recordings they play fine.  Therefore I installed leanfront to see if it acted the same way.  I am able to connect to the ver. 32 backend with leanfront and I liked it, in fact I think the video is better than with Kodi.  When the damaged recordings are marked with yellow lettering in Kodi they show up in leanfront looking like a file folder rather than the normal program icons.  When I try to open or play these recordings that look like a file folder they will not play and give a message that the media source is bad.  I think this may be a bug with mythtv ver. 32 as far as marking the damaged recordings, but I was wondering if something can be done to make these recordings with this issue play.

Number two:  I do not seem to be able to enable commercial skipping in leanfront.  I really like this feature in Kodi and I would be inclined to use leanfront if commercial skipping could be enabled.  I looked on line and I found a post that commercial skipping is simply not available on leanfront and the arrow keys would have to be used to advance through commercials on recordings.  Is commercial skipping something that is being looked at with leanfront?

**Comment by bennettpeter** [comment]
There was a change in the flag values for marking a recording as bad between V31 and V32. I made a change to leanfront to cater for this, so that recordings would not be shown as damaged incorrectly. Perhaps the Kodi plugin did not make that change.

I don't understand why recordings would show as a file folder. If there is no artwork for the recording it will show as a movie icon, a rectangle with three grooves in the the top, supposed to look like a clapperboard.

Please send me an extract of a video that will not play, so that I can investigate it and possibly get a fix or recommend a setting that will work. You can share a link to google drive or similar service where you can upload the video.

leanfront supports forward and back skips using the arrow keys or ff/rew keys. This is what I use for skipping.

leanfront does not support skipping using commercial break markers in MythTV. I assume that is what you are referring to. I have never used commercial detection. The times I have tried it have not been successful. What settings do you use for commercial detection and how well does it work? If it actually works correctly for somebody I can look into supporting it in leanfront. My impression has been that few people actually use it and get any benefit. Perhaps you mark the commercials manually using mythfrontend?

**Comment by bennettpeter** [comment]
There is a fix in the works for the kodi plugin for the damaged indicator. see https://github.com/janbar/pvr.mythtv/issues/197 .

**Comment by bennettpeter** [comment]
If you have a google logon or a gmail account you get a google drive. Go to google.com and you will either see "sign in" at the top right hand side or you will see your picture if you are signed in. If you click "sign in" you can create an account if you don't have one. Once you have created an account, go to google.com, click on the "google apps" icon (9 dots) at the top right hand side. Select "drive". Now you can create directories, drop files, and share the files. You get 15 GB of storage to put whatever you want.

**Comment by rocknrobin72** [comment]
Okay, I can do that as far as the google drive is concerned.  My next question is how do I pull out an extract of an affected video.  As I mentioned I had one yesterday that showed the yellow lettering in Kodi, but showed the correct artwork in leanfront instead of the movie icon and it played okay.  I could post that one, but I don't think that will point out what I have been seeing since leanfront played the recording okay.

**Comment by bennettpeter** [comment]
I only need to see recordings that will not play in leanfront but play in other programs, such as mythfrontend or vlc.
To find the recording file you can press I (for information) in the recording list when the recording is selected. Among the fields displayed is "Recorded File Name". It will be something like 3763_20161107030000.ts. Search for the file in your Default Storage Group directories.
If the file is too big to upload, extract 100MB with dd. Example:
`dd bs=1MB count=100 skip=100 if=3763_20161107030000.tsx of=extracted_with_dd.ts`
This example will skip the first 100MB then extract the next 100MB. If you want the first 100MB leave out the "skip=100".

**Comment by rocknrobin72** [comment]
After quite a bit more evaluation I seen that leanfront seems to be working okay in conjunction with MythTV ver. 32 as far as damaged recordings are concerned.  When I see recordings in Kodi with the mythtv pvr add-on and they play they play ok in leanfront.  Evidently the Kodi version of mythtv add-on is not looking at the right flag for damaged recordings because even when the lettering is in yellow they normally play ok and of course play ok in leanfront.  I have also seen where the recording appears ok in Kodi, as far as the lettering is concerned, but then leanfront will show the movie icon as you mentioned with a little red and yellow circle, like a little sun, in the lower left corner and then the recording will not play in either Kodi or leanfront.  Then sometimes the recording will show okay in Kodi, but shows the artwork in leanfront with a little red sun in the lower left corner.  In this case the recording plays on both Kodi and leanfront, but may have short periods where it's all pixelated and audio is distorted.   I am recording from an off the air antenna with hdhomerun tuners so I assume these are periods when the signal is not good.  Therefore I believe leanfront is working properly displaying the damaged recordings.   

If you had the time I would really like to see commercial skips working.  Although they are not perfect, as I mentioned the mythcommflag algorithm either misses a commercial period sometimes or is too aggressive and marks commercials all the way to the end of a show, I sure like the commercial skips when they work.

**Comment by bennettpeter** [comment]
The little "explosion" icon in leanfront is the indicator for a damaged recording and corresponds to the yellow outline in mythfrontend, so that seems correct.

I will leave this open as an enhancement request for commercial skip. I don't know if or when I will get around to it. If there were  more people requesting it, I may be more motivated to do it.

**Comment by rocknrobin72** [comment]
Thanks for your responses and help.  I find myself using leanfront as an alternative to kodi because, as I've mentioned, the video seems to be crisper and is smoother sometimes.  Leaving the commercial skip as an enhancement request is fine with me.  I can live with it as is, but it would be nice...

**Comment by bennettpeter** [comment]
I am ready to add commercial-skip to leanfront. I ran a commercil csna on a recording and it did a pretty good job detecting the commercials, although it missed at least one. 

One thing - it correctly identified the beginning of the recording as a commercial break and flagged it but mythfrontend is unable to skip a commercial break that starts at position 0. I will fix that with leanfront. 

As far as commercials at the end of the show are concerned - leanfront does **not** exit at the end of the recording. It stops playback and you can press back arrow to play the last part again. That should help with your problem at the end of a show. I find the exit at end of playback to be very irritating, I often want to go back a bit and see the last part if I accidentally skip past it.

I see that with MythTV, going forward or backward with the arrows does not skip commercials, only playing into them skips. That is probably why the commercial at the beginning of the show is not skipped, even though it is marked.

Before I start on this - Are you still looking for the commercial skip feature?

**Comment by rocknrobin72** [comment]
Yes Sir Peter,

 

I would love to see the commercial skips incorporated in the leanfront.  I have been using it since you suggesting sideloading it to my Android TV and I find it works very well as a frontend.  

 

Even though the commercial skips aren’t perfect, may miss one every now and then, and can get aggressive at the end of the show (and I realize these are all due to the algorithm in the backend that marks the commercial skips, not the frontend) I still like having that feature there.  I assume you will have the option, as in the MythTV frontend and Kodi, to turn the feature on and off?   Some folks may not want to use it.

 

Thanks.  

 

Rob Rutledge

**Comment by rocknrobin72** [comment]
I tried the new trial version as soon as I saw this email.  I saw in the settings that I could enable the commercial skips automatically which I did.  When I tried to play the first program it started and then “bombed off”; stopped the program and went back to the window to start the program again.  I deleted the new trial version and re-installed the older version without the commercial skips and it played the program okay again.  I will re-install the new trial version tomorrow and not enable the commercial skips and see if it works okay that way.  

 

Rob Rutledge

**Comment by bennettpeter** [comment]
I have fixed the bug that was causing a crash on MythTV V31. Download today's new version.

**Comment by rocknrobin72** [comment]
Peter,

 

Thanks, that seemed to do the trick.  After viewing some of the programs I have the following observations:

 

*	When I was able to play back the first program I experienced a skipping  in the audio and video and the little circle showed up that I assume signifies buffering.  This would have a few times and then the program would stop playback.  I could use the play button to start it again and then it would begin the skipping, hiccupping in audio and video and then stopping.  After several times of restarting it stopped and I got this message “Playback Failed.  Error occurred in Renderer.  MediaCodecAudioRenderer error, index =1, format=FORMAT(1/84,null,null,audio/ac3,null,-1,en[-1-1,-1.0],[2,48000]),format_supported=YES    EXIT   CONTINUE”.  If I exited it went back to the screen to start the program again.  If I clicked on Continue the Program would begin again with the audio/video stuttering and stopping and ultimately giving me the same message.   I continued to click on Continue until after about the fourth or fifth time the program finally started playing okay with no stuttering.  I am assuming that the program was trying to figure out which transcoder to use account under settings, Playback:Default there was a setting for Automatic, ffmpeg, and another option I can’t remember.  Now the setting for Automatic and the other three settings is not there.  I am assuming, once again, that the program figured out which transcoder to use, again I am assuming that it eventually decided on ffmpeg and the other transcoder did not work right.  I pass this along to you as FYI account maybe you understand what transpired there.  Additionally after the first program I watched, and I have now watched three other programs, this behavior did not re-occur.  Every program starts out playing fine now.  
*	Once I got over the hurdle mentioned above the program(s) would play okay and the commercial skipping worked as well as the commercial skip works on Kodi.  I did notice sometimes right as the skip happened the buffering circle would appear briefly which is no big deal.  It did seem that after the skip happened sometimes there was a momentary lapse in the audio as it possibly had to synch up again or something.  These are just minor inconveniences, not show stoppers so I liked how the commercial skipping did work.  

 

I would like to ask for one improvement if it’s not a big deal.  Is it possible to display the “tics” on the timeline when the program is playing that signify when a commercial skip is going to happen?  These appear as red areas or blocks on the timeline in Kodi and I believe the standard MythTV frontend shows them as well.  These are handy to illustrate when the skipping is going to happen and with that in mind where the commercial skips algorithm may have missed commercial skips or gotten too aggressive and blocked out some of the program.  

 

All in all I like what I see so far.

 

Thanks.    

 

Rob Rutledge
